### PR TITLE
feature/loader

### LINKED ISSUE #25: Should have a loader
It would be nice to have a simple loader that appears during initialization.

### PR DESCRIPTION
**Pull request by joshua-dean**
# Loader
Closes #25.

## Description
- Added overlay with spinning loading icon 
    - Added during init and removed once loading is complete
- Pulled a lot of init logic into it's own file
- Added a `staggered_ulabel_init` which displays the screen earlier and adds a configurable delay between each step. This is intended to help visualize/debug the loading process, and demonstrate features like the loader 
    - This branch uses the "staggered" init with a 250ms delay per step. This is currently hard-coded but we could make this a feature flag or configurable. 
- Added `build-dev-and-demo` npm script
- Added a bunch of missing init functions to ULabel's type declaration 

The "staggered" init is `async` to allow the delays to run. The `ULabel.load_image_promise` wasn't playing nicely with `async`, so I opted to use `.decode()` instead (why wasn't this done in the first place?). 

I resisted the urge to completely rearrange and init process in this PR, but now there is a place to do so if we want to optimize in the future.

~The production build for this is way slower than previously, which I suspect is due to the static function calls (`index.js` and `initializer.ts` both call each other so there's some resolution that webpack has to handle). I'm going to leave this as a draft until I can solve this, or we decide that we don't care.~
Production build size issues solved by #200 and #201.

## PR Checklist

- [X] Merged latest ~main~ target branch
- [ ] Version number in `package.json` has been bumped since last release
- [ ] Version numbers match between package `package.json` and `src/version.js`
- [ ] Ran `npm install` and `npm run build` AFTER bumping the version number
- [ ] Updated documentation if necessary (currently just in `api_spec.md`)
- [ ] Added changes to `changelog.md`

## Breaking API Changes
maybe

### CODE REVIEW

**Comment by joshua-dean** [comment]
Yeah so the really long builds is a result of webpack increasing the size of `dist/` every build, because `initializer.ts` calls static ULabel functions in `index.js`, so it's a somewhat circular dependency. It's possible there's a setting to make webpack not do this. 

To avoid this, we would need to pull `has_night_mode_cookie`, `initialize_annotation_canvases`, `load_image_promise`, `create_listeners` (massive), `after_init`, and `elvl_fatal` to not be in `index.js`.

Static class functions that take an instance of the class seem problematic in our use case anyways, so hoisting these elsewhere is probably in our best interest. I might punt on this specific feature and do some PRs on the other ones (like `create_listeners`) separately.

**Comment by joshua-dean** [comment]
> I might punt on this specific feature and do some PRs on the other ones (like `create_listeners`) separately.

The necessary changes to avoid the issue are in #200 and #201. I have merged the latest of #201 into this branch and made it the base for this PR. I have tested this and am able to build repeatedly without `dist/` ballooning in size.

**Comment by joshua-dean** [comment]
#200 and #201 are approved, so I merged the latest and am marking this as "ready for review".

File: index.d.ts [hunk 1]
```diff
@@ -266,8 +268,14 @@ export class ULabel {
...
     static get_allowed_toolbox_item_enum(): AllowedToolboxItem;
-    static process_classes(ulabel_obj: ULabel, arg1: string, subtask_obj: ULabelSubtask);
-    static build_id_dialogs(ulabel_obj: ULabel);
+    static process_classes(ulabel_obj: ULabel, arg1: string, subtask_obj: ULabelSubtask): void;
+    static build_id_dialogs(ulabel_obj: ULabel): void;
+
+    // Instance init functions
+    public create_overlays(): void;
+
+    // nops
```
Comment (TrevorBurgoyne) [inline index.d.ts:277]: i may be stupid, what is `nops`?

File: src/loader.ts [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+            }
+            
+            @keyframes spin {
+                0% { transform: rotate(0deg); }
+                100% { transform: rotate(360deg); }
+            }
+        `;
+        const style = document.createElement('style');
+        style.innerHTML = css;
+        document.head.appendChild(style);
```
Comment (TrevorBurgoyne) [inline src/loader.ts:57]: on the one hand i like having css close to the thing its actually styling, on the other is there a reason to attach these styles to the document and have them persist after the loader div is removed? versus we could attach the styles directly to the elements rather than making global class definitions

File: src/index.js [hunk 3]
```diff
@@ -636,112 +629,8 @@ export class ULabel {
...
-                LogLevel.ERROR,
-            );
-        });
-
-        // Final code to be called after the object is initialized
-        this.after_init();
-
-        console.log(`Time taken to construct and initialize: ${Date.now() - this.begining_time}`);
+        // ulabel_init(this, callback);
+        staggered_ulabel_init(this, callback, 250);
```
Comment (TrevorBurgoyne) [inline src/index.js:633]: any reason to keep the `ulabel_init` method? only thing i can think of is if we want to expose an option in the config to disable the loader

**Review by TrevorBurgoyne — COMMENTED** [review]
loader is very cool, and i like the general init refactor

File: index.d.ts (hunk 1 above)
Comment (joshua-dean) [reply index.d.ts:277]: https://en.wikipedia.org/wiki/NOP_(code)

tl;dr `redraw_demo()` doesn't actually do anything

File: src/loader.ts (hunk 2 above)
Comment (joshua-dean) [reply src/loader.ts:56]: Yeah we could keep track of the `<style>` element and also remove that, or maybe make it a child of the overlay?
The latter is the easiest (I think), just change `add_loader_css()` to `build_style()` that returns the `<style>` element, and let `add_loader_div()` handle where to attach it.

File: src/index.js (hunk 3 above)
Comment (joshua-dean) [reply src/index.js:633]: The delays are really only for visualization, but we could consolidate to a single method that just sets the delays to 0 in the default case. 
We could also consolidate to one method without delays. 

Alternative to all of the above, we could have a minimum fixed loading time to always show off the loader, picking and choosing which components visually render in behind the overlay. There is some validity to artificial delays for UX (see [this](https://90percentofeverything.com/2010/12/16/adding-delays-to-increase-perceived-value-does-it-work/)), but I definitely would want that to be configurable - there is at least one ULabel user that would prefer instantaneous loading (me).

File: src/loader.ts (hunk 2 above)
Comment (joshua-dean) [reply src/loader.ts]: I went ahead and did this in 5480b19, and confirmed that the style got removed. I like this structure better anyways.

Also in the previous version I was adding the style twice, which was probably harmless but not ideal.

File: src/index.js (hunk 3 above)
Comment (TrevorBurgoyne) [reply src/index.js:633]: yeah i dont think we need to add an option for a fixed loading time. one thought i had is that perhaps we could support certain config args to be Promises and then we can await them in the init so that the loader is more versatile. IE if a user is hitting an API to retrieve an image url, they could pass a promise to ulabel and then the loader could start before the image promise resolves. thoughts?

Comment (joshua-dean) [reply src/index.js:633]: Do you think for now I should keep the async one, bonk the old one and all the delays?

Comment (TrevorBurgoyne) [reply src/index.js:633]: if the delays were just for demoing the loader then for now we can remove them all or use the old init. then we should look into allowing users to choose when to start the loader, like maybe we provide a static method that allows the user to start the loader before even calling the ulabel constructor, and that same method that runs in the html_bulider would check if a loader is already present before adding one

Comment (TrevorBurgoyne) [reply src/index.js:633]: although if the ulabel html and styles arent present yet that might not work actually

Comment (joshua-dean) [reply src/index.js:633]: Yeah what I noticed while testing this is that the ULabel container is intentionally not visible for a good portion of the loading process, so in `staggered_ulabel_init` I moved the step that makes it visible to be way earlier to show what's actually going on. 

We could just tweak the container hierarchy slightly so that the loader is attached to the main div immediately, and everything else is attached to another div that waits a bit before displaying. That would be set up in a way to allow it to be launched earlier if desired. 

Another consideration is how the `.on` callback utility works with async functions. Consider:
```javascript
async function fn() {
...
}
ulabel.on(ulabel.fn, () => { ... })
// Does this await just fn, or also it's attached callbacks?
await ulabel.fn();
```
Easy enough to test, and if it doesn't it should be possible to add that functionality through a different method. 
If we got that working, we could just structure initialization steps together and let the user attaching loading logic wherever they want, including when to display the loader.

Comment (TrevorBurgoyne) [reply src/index.js:633]: yeah after thinking about it a bit, i think in my head i want two uses for the loader. First is what is implemented here, which is a loader that displays as ulabel performs its init. Second is a set of static methods so that a user can choose when to show the same style of loader while they are doing whatever prep is needed before calling ulabel's constructor and init, like loading the image and annotations etc. We could split the second way off into its own issue and release what we have here for now.

Comment (joshua-dean) [reply src/index.js:633]: For the most part this could use the same logic. Taking inspiration from [Textual](https://textual.textualize.io/), where every widget has a reactive `.loading` attribute which will add a loading icon and an overlay, we could do something similar. We would just have to refactor the logic to be entirely separate from the ULabel constructor, but also called as part of the normal `init` logic.

Comment (joshua-dean) [reply src/index.js:633]: Opened #206 to track this.

**Comment by joshua-dean** [comment]
@TrevorBurgoyne I consolidated to a single init (the async one) without any delays, and merged the latest main through the whole chain.

#200 and #201 are approved, so if this is approved I'll merge them all into #202 so we can release them all together.

**Review by TrevorBurgoyne — APPROVED** [review]
LGTM 👍
